xds/internal/client/client_xds.go

```diff
@@ -53,52 +58,137 @@ func UnmarshalListener(resources []*anypb.Any, logger *grpclog.PrefixLogger) (ma
+	// server-side as in the following format:
+	// grpc/server?udpa.resource.listening_address=IP:Port.
+	addr := lis.GetAddress()
+	if addr == nil {
+		return nil, fmt.Errorf("xds: no address field in LDS response: %+v", lis)
+	}
+	sockAddr := addr.GetSocketAddress()
+	if sockAddr == nil {
+		return nil, fmt.Errorf("xds: no socket_address field in LDS response: %+v", lis)
+	}
+	host, port, err := getAddressFromName(lis.GetName())
+	if err != nil {
+		return nil, fmt.Errorf("xds: no host:port in name field of LDS response: %+v", lis)
+	}
+	if h := sockAddr.GetAddress(); host != h {
+		return nil, fmt.Errorf("xds: socket_address host does not match the one in name. Got %q, want %q", h, host)
+	}
+	if p := strconv.Itoa(int(sockAddr.GetPortValue())); port != p {
+		return nil, fmt.Errorf("xds: socket_address port does not match the one in name. Got %q, want %q", p, port)
+	}
+
+	// Make sure that the listener resource contains a single filter chain with
+	// the application_protocols field set to "managed-mtls", and the
+	// "transport_socket" field containing the appropriate security
+	// configuration.
+	fcs := lis.GetFilterChains()
+	if n := len(fcs); n != 1 {
+		return nil, fmt.Errorf("xds: filter chains count in LDS response does not match expected. Got %d, want 1", n)
+	}
+	fc := fcs[0]
+	aps := fc.GetFilterChainMatch().GetApplicationProtocols()
+	if len(aps) != 1 || aps[0] != "managed-mtls" {
+		return nil, fmt.Errorf("xds: application_protocols in LDS response does not match expected. Got %v, want %q", aps, "managed-mtls")
+	}
+
+	// If the transport_socket field is not specified, it means that the control
+	// plane has not sent us any security config. This is fine and the server
+	// will use the fallback credentials configured as part of the
+	// xdsCredentials.
+	ts := fc.GetTransportSocket()
+	if ts == nil {
+		return &ListenerUpdate{}, nil
+	}
+	if name := ts.GetName(); name != transportSocketName {
+		return nil, fmt.Errorf("xds: transport_socket field has unexpected name: %s", name)
+	}
+	any := ts.GetTypedConfig()
+	if any == nil || any.TypeUrl != version.V3DownstreamTLSContextURL {
+		return nil, fmt.Errorf("xds: transport_socket field has unexpected typeURL: %s", any.TypeUrl)
+	}
+	downstreamCtx := &v3tlspb.DownstreamTlsContext{}
+	if err := proto.Unmarshal(any.GetValue(), downstreamCtx); err != nil {
+		return nil, fmt.Errorf("xds: failed to unmarshal DownstreamTlsContext in LDS response: %v", err)
+	}
+	if downstreamCtx.GetCommonTlsContext() == nil {
+		return nil, errors.New("xds: DownstreamTlsContext in LDS response does not contain a CommonTlsContext")
+	}
+	sc, err := securityConfigFromCommonTLSContext(downstreamCtx.GetCommonTlsContext())
+	if err != nil {
+		return nil, err
+	}
+	return &ListenerUpdate{SecurityCfg: sc}, nil
+}
+
+func getAddressFromName(name string) (host string, port string, err error) {
+	var addr string
+	_, err = fmt.Sscanf(name, "grpc/server?udpa.resource.listening_address=%s", &addr)
```

Comment:
Didn't #4049 change this?

Reply:
Thanks for the catch. I have fixed it now. But it looks a little gross. I couldn't get this to work with the regexp package. Do you have any suggestions?